mobile/library/cc/stream_callbacks.cc

```diff
@@ -118,7 +118,7 @@ envoy_http_callbacks StreamCallbacks::asEnvoyHttpCallbacks() {
       &c_on_complete,
       &c_on_cancel,
       &c_on_send_window_available,
-      new StreamCallbacksSharedPtr(shared_from_this()),
+      shared_from_this(),
```

Comment:
I think you can do this `static_cast<void*>(shared_from_this().get())` and then update the callbacks in https://github.com/envoyproxy/envoy/blob/ffbf1889ef788ccb4b1d0e402ed2c3976d7204fd/mobile/library/cc/stream_callbacks.cc#L13-L107 to `static_cast<StreamCallbacks*>(context)`. Afterwards, we can then remove the manual `delete` since the pointer is now backed by `shared_ptr`.

Reply:
You can't do this, because if you get shared_from_this(), then get a raw pointer to it, the object retrieved from shared_from_this() will go out of scope when the function exits, and if the StreamPrototype goes out of scope before the `Stream` does (which is entirely plausible), then the underlying memory will get reclaimed before the raw pointer is used in the StreamCallbacks callback functions.

I verified this actually by changing the FetchClient to return the Stream from StreamPrototype, then have the StreamPrototype go out of scope while using the Stream, and got a crash error with `asan`.